### PR TITLE
[2.6.3] Rip out etcd binary

### PR DESCRIPTION
**Pull request by annablender**
Issue [33850](https://github.com/rancher/rancher/issues/33850). Exploratory issue that removes the downloaded etcd binary and starts Rancher single node k3s cluster with embedded etcd.

#### Design plan

- Rip etcd binary and env vars out of rancher/rancher Dockerfile.dapper and package/Dockerfile
- Update the Norman API
    - Remove handler function RunETCD that starts etcd
    - Remove —datastore-endpoint
    - Start k3s server with —cluster-init (which will use embedded etcd)

#### What was fixed, or what changes have occurred

- Remove etcd binary and env vars
- Update K3S_URL in Dockerfile to avoid overwriting k3s server url during setup
- Update go.mod to reference changes in Norman API (the k3s server initialization is done in the Norman API and that's where the other half of these changes are)
- Fix bug in the validate script
- Add missing /var/lib/rancher directory

#### Areas or cases that should be tested

- Spin up a custom Rancher image v2.6 and test if 1) the local cluster is stable, 2) a deployment can be created, 3) a downstream cluster can be created. I used a custom image modified for local dev that can be pulled from docker hub _annablender/rancher:k3s-embedded-etcd_.
- Drop into the Rancher container and ping the embedded etcd endpoint "curl -i localhost:2379." If running, you'll establish a connection and receive empty data
- K3s single node upgrade
- Persisted data upgrade
- Backup and restore
- Use etcdctl with flag --endpoints=https://<node ip>:2379 to test embedded etcd

### CODE REVIEW

**Review by brandond — COMMENTED** [review]
LGTM once the norman PR is merged and go.mod updated.

**Comment by nickgerace** [comment]
Will re-review once https://github.com/rancher/norman/pull/406 is merged and the module is updated for this PR

File: package/entrypoint.sh
Comment (brandond) [inline package/entrypoint.sh:14]:
```suggestion
fi
if [ -e /var/lib/rancher/k3s/server/db/etcd ]; then
  k3s server --cluster-reset
fi
```

File: gomodtidy.sh
Comment (brandond) [inline gomodtidy.sh:1]: not sure about this one

**Review by brandond — CHANGES_REQUESTED** [review]

File: package/entrypoint.sh
Comment (annablender) [reply package/entrypoint.sh:14]: Ahh good point

File: gomodtidy.sh
Comment (annablender) [reply gomodtidy.sh:1]: Oh that's a script I wrote for updating go modules! My bad, I'll remove that

File: package/entrypoint.sh
Comment (brandond) [inline package/entrypoint.sh:15]: Note for other reviewers: this is necessary because K3s's embedded etcd can't be bound to 127.0.0.1, and will get mad if the Docker container IP address changes. This `--cluster-reset` doesn't drop any data from the etcd datastore; it just resets the cluster membership list and ensures that the sole etcd cluster member has the correct IP address.

**Review by brandond — APPROVED** [review]

**Review by nickgerace — DISMISSED** [review]

**Comment by ibuildthecloud** [comment]
R.I.P.

**Review by nickgerace — DISMISSED** [review]

**Comment by brandond** [comment]
lets get the conflicts resolved and :ship::it:!

File: Dockerfile.dapper542220813
Comment (superseb) [inline Dockerfile.dapper542220813:1]: Can we remove file `Dockerfile.dapper542220813`?

**Comment by superseb** [comment]
@annablender is `annablender/rancher:k3s-embedded-etcd` still the image to test with?

File: package/entrypoint.sh
Comment (superseb) [inline package/entrypoint.sh:15]: What is the reason to suppress the log here? If this fails, the container log is empty and the user does not know what's going on.

And do we need any counter to when this fails to recover automatically or if this fails, it needs manual intervention and we need to point that out? Or is that covered when we show the log from `--cluster-reset`?

Comment (brandond) [reply package/entrypoint.sh:15]: The log from this operation is quite noisy, and will not normally be something that the user needs to see. We could probably check to see if it succeeds, and cat the output if it fails?

Comment (annablender) [reply package/entrypoint.sh:15]: @superseb Is this ok with you?

Comment (superseb) [reply package/entrypoint.sh:15]: Yeah we dont need more noisy stuff in there but we can't not let the user know whats going on in this case. Exit code check and showing the log is fine.

Does a `cluster-reset` failing for some reason always require manual intervention? So we don't need some sort of auto recovery in that case and just always manual intervention where the user can use the printed log? I'm asking this in relation to how it was, and that was k3s always trying to start (it didn't do a cluster reset afaik)

Comment (brandond) [reply package/entrypoint.sh:15]: The cluster-reset isn't really for k3s, it's for etcd. Since we can't tell k3s to use the loopback address for the cluster member IP, we need to reset the cluster member address to the current container address every time, or etcd will refuse to start.

https://github.com/rancher/rancher/pull/35323#discussion_r747068862

**Review by nickgerace — APPROVED** [review]

File: package/entrypoint.sh
Comment (brandond) [inline package/entrypoint.sh:16]: I'm pretty sure you'll need whitespace around this
```suggestion
  if [ $? -ne 0 ]; then
```

**Review by brandond — CHANGES_REQUESTED** [review]

**Review by superseb — APPROVED** [review]

**Review by nickgerace — APPROVED** [review]

**Comment by nickgerace** [comment]
If/once @brandond approves, I'll merge.

**Review by brandond — APPROVED** [review]

**Pull request merged**
